pybandits/smab_simulator.py

```diff
@@ -0,0 +1,139 @@
+# MIT License
+#
+# Copyright (c) 2022 Playtika Ltd.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import random
+from typing import Dict, List, Tuple
+
+import numpy as np
+import pandas as pd
+
+from pybandits.base import ActionId, BinaryReward
+from pybandits.pydantic_version_compatibility import Field, model_validator
+from pybandits.simulator import Simulator
+from pybandits.smab import BaseSmabBernoulli
+
+
+class SmabSimulator(Simulator):
+    """
+    Simulate environment for stochastic multi-armed bandits.
+
+    This class performs simulation of stochastic Multi-Armed Bandits (sMAB). Data are processed in batches of size n>=1.
+    Per each batch of simulated samples, the mab selects one a1nd collects the corresponding simulated reward for
+    each sample. Then, prior parameters are updated based on returned rewards from recommended actions.
+
+    Parameters
+    ----------
+    mab : BaseSmabBernoulli
+        sMAB model.
+    """
+
+    mab: BaseSmabBernoulli = Field(validation_alias="smab")
+    _base_columns: List[str] = ["batch", "action", "reward"]
+
+    @model_validator(mode="before")
+    @classmethod
+    def replace_null_and_validate_probs_reward(cls, values):
```

Comment:
Can't this be handled directly by pydantic?
You can set a default value directly in the definition. If the input is None, Pydantic will use the default value.
E.g. (with int)
```
x: Optional[int] = 13  # Default value is 13 (static definition)
x: Optional[int] = Field(default_factory=lambda: 13)  # Default value is 13 (dynamic definition)
```

Reply:
probs_reward number of columns is determined from the number of actions, so I cannot use your suggested dynamic definition. Agree?